src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,521 @@
+	flushWorkers.Init()
+	handler := newDownsamplerFlushHandler(o.Storage, sortedTagIteratorPool,
+		flushWorkers, instrumentOpts)
+	aggregatorOpts = aggregatorOpts.SetFlushHandler(handler)
+
+	return newAggregatorResult{
+		aggregator:              aggregator.NewAggregator(aggregatorOpts),
+		matcher:                 matcher,
+		tagEncoderPool:          tagEncoderPool,
+		tagDecoderPool:          tagDecoderPool,
+		encodedTagsIteratorPool: sortedTagIteratorPool,
+	}, nil
+}
+
+type downsamplerFlushHandler struct {
+	sync.RWMutex
+	storage                 storage.Storage
+	encodedTagsIteratorPool *encodedTagsIteratorPool
+	workerPool              xsync.WorkerPool
+	instrumentOpts          instrument.Options
+	metrics                 downsamplerFlushHandlerMetrics
+}
+
+type downsamplerFlushHandlerMetrics struct {
+	flushSuccess tally.Counter
+	flushErrors  tally.Counter
+}
+
+func newDownsamplerFlushHandlerMetrics(
+	scope tally.Scope,
+) downsamplerFlushHandlerMetrics {
+	return downsamplerFlushHandlerMetrics{
+		flushSuccess: scope.Counter("flush-success"),
+		flushErrors:  scope.Counter("flush-errors"),
+	}
+}
+
+func newDownsamplerFlushHandler(
+	storage storage.Storage,
+	encodedTagsIteratorPool *encodedTagsIteratorPool,
+	workerPool xsync.WorkerPool,
+	instrumentOpts instrument.Options,
+) handler.Handler {
+	scope := instrumentOpts.MetricsScope().SubScope("downsampler-flush-handler")
+	return &downsamplerFlushHandler{
+		storage:                 storage,
+		encodedTagsIteratorPool: encodedTagsIteratorPool,
+		workerPool:              workerPool,
+		instrumentOpts:          instrumentOpts,
+		metrics:                 newDownsamplerFlushHandlerMetrics(scope),
+	}
+}
+
+func (h *downsamplerFlushHandler) NewWriter(
+	scope tally.Scope,
+) (writer.Writer, error) {
+	return &downsamplerFlushHandlerWriter{
+		ctx:     context.Background(),
+		handler: h,
+	}, nil
+}
+
+func (h *downsamplerFlushHandler) Close() {
+}
+
+type downsamplerFlushHandlerWriter struct {
+	wg      sync.WaitGroup
+	ctx     context.Context
+	handler *downsamplerFlushHandler
+}
+
+func (w *downsamplerFlushHandlerWriter) Write(
+	mp aggregated.ChunkedMetricWithStoragePolicy,
+) error {
+	w.wg.Add(1)
+	w.handler.workerPool.Go(func() {
+		defer w.wg.Done()
+
+		logger := w.handler.instrumentOpts.Logger()
+
+		iter := w.handler.encodedTagsIteratorPool.Get()
+		iter.Reset(mp.ChunkedID.Data)
+		expected := iter.tagDecoder.Remaining()
+		if len(mp.ChunkedID.Suffix) > 0 {
+			expected++
+		}
+		tags := make(models.Tags, expected)
+		for iter.Next() {
+			name, value := iter.Current()
+			tags[string(name)] = string(value)
+		}
+		if len(mp.ChunkedID.Suffix) > 0 {
+			tags[aggregationSuffixTag] = string(mp.ChunkedID.Suffix)
+		}
+
+		err := iter.Err()
+		iter.Close()
+		if err != nil {
+			logger.Debugf("downsampler flush error preparing write: %v", err)
+			w.handler.metrics.flushErrors.Inc(1)
+			return
+		}
+
+		err = w.handler.storage.Write(w.ctx, &storage.WriteQuery{
```

Comment:
I believe for this we'll allow the M3DB session client to retry the write (it has a configurable write retrier), just so there is one way to do it.  We do it at a higher level in the ingester but only because we want to know when we're performing retries, etc with metrics.